### ISSUE TITLE
MSYS2 6.3.0 build fails

### ISSUE THREAD

**Issue by binji** [WebAssembly/wabt#324]
The MSYS2 Mingw64 build fails here. Looks like it doesn't like `PRIu64`

```
C:/projects/wabt/src/validator.cc: In function 'void check_limits(Context*, const WabtLocation*, const WabtLimits*, uint64_t, const char*)':
C:/projects/wabt/src/validator.cc:627:52: error: unknown conversion type character 'l' in format [-Werror=format=]
                 desc, limits->initial, absolute_max);
                                                    ^
C:/projects/wabt/src/validator.cc:627:52: error: unknown conversion type character 'l' in format [-Werror=format=]
C:/projects/wabt/src/validator.cc:627:52: error: too many arguments for format [-Werror=format-extra-args]
C:/projects/wabt/src/validator.cc:633:50: error: unknown conversion type character 'l' in format [-Werror=format=]
                   desc, limits->max, absolute_max);
                                                  ^
C:/projects/wabt/src/validator.cc:633:50: error: unknown conversion type character 'l' in format [-Werror=format=]
C:/projects/wabt/src/validator.cc:633:50: error: too many arguments for format [-Werror=format-extra-args]
C:/projects/wabt/src/validator.cc:639:59: error: unknown conversion type character 'l' in format [-Werror=format=]
                   desc, limits->max, desc, limits->initial);
                                                           ^
C:/projects/wabt/src/validator.cc:639:59: error: format '%s' expects argument of type 'char*', but argument 5 has type 'uint64_t {aka long long unsigned int}' [-Werror=format=]
C:/projects/wabt/src/validator.cc:639:59: error: unknown conversion type character 'l' in format [-Werror=format=]
C:/projects/wabt/src/validator.cc:639:59: error: too many arguments for format [-Werror=format-extra-args]
cc1plus.exe: all warnings being treated as errors
```

**Comment by rongjiecomputer** [comment]
I added this to `config.h`

```cpp
#ifdef __MINGW32__
#include <inttypes.h>

#undef PRIu64
#undef PRId64
#undef PRIx64
#define PRIu64 "I64u"
#define PRId64 "I64d"
#define PRIx64 "I64x"

#endif
```

Checking the preprocessed `src\binary-reader.cc` with `-E`, `PRIu64` is replaced with `I64u` correctly ([see file](https://gist.github.com/rongjiecomputer/06b2021c7a1dce374366a4fef16a1382#file-binary-reader-cc-L3425)), but still Mingw64 complains:

```
src/binary-reader.cc: In function 'void sprint_limits(char*, size_t, const WabtLimits*)':
src/binary-reader.cc:644:56: warning: format '%u' expects argument of type 'unsigned int', but argument 4 has type 'uint64_t {aka long long unsigned int}' [-Wformat=]
                            limits->initial, limits->max);
                                                        ^
```

I have no idea where this `%u` comes from.

And this is just one of the many warnings that I don't see in Linux GCC. [Complete list of warning](https://gist.github.com/rongjiecomputer/06b2021c7a1dce374366a4fef16a1382#file-warning-txt).

**Comment by binji** [comment]
Thanks for looking into this! It seems very strange to me. Just to confirm, you are using MSYS2 6.3.0?

**Comment by rongjiecomputer** [comment]
I am using Mingw64 GCC 6.3.0 compiled from source myself rather than downloading from sourceforge, but I make no changes to the source, so bugs exist in sourceforge distro will appear in mine too.

**Comment by binji** [comment]
Sorry, I'm still a bit confused. The gist you link to above (labeled "see file"), is that generated from your local MSYS build, or were you able to extract it from the appveyor builder? The reason I ask is that your PR #327 says it compiles on your machine, but that gist makes it seem like it's breaking in the same way.

Thanks again for taking a look, I've been too lazy to fire up my Windows partition to check myself :-)

**Comment by rongjiecomputer** [comment]
Before I made PR #327, I use custom build script with `-Werror` disabled, saw a bunch of warnings but not converted to errors. I only realized that I missed this flag locally after checking failed appveyor log, so I closed the PR. Should have explained about this before closing the PR.

And yes, my gist was generated from my local machine's Mingw64 6.3.0.

**Comment by rongjiecomputer** [comment]
I produced this small test case and compiled with `g++ -o test test.cpp -Wall -Wextra`, no warning.

```cpp
#define __STDC_LIMIT_MACROS 1
#define __STDC_FORMAT_MACROS 1
#include <stdio.h>
#include <stdint.h>
#include <inttypes.h>

struct WabtLimits {
  uint64_t initial;
  uint64_t max;
  bool has_max;
};

#define wabt_snprintf snprintf

static void sprint_limits(char* dst, size_t size, const WabtLimits* limits) {
  int result;
  if (limits->has_max) {
    result = snprintf(dst, size, "initial: %" PRIu64 ", max: %" PRIu64,
                           limits->initial, limits->max);
  } else {
    result = snprintf(dst, size, "initial: %" PRIu64, limits->initial);
  }
  static_cast<void>(result);
}

int main() {
	size_t a = 1llu << 63;
	printf("%I64u\n", a);
	printf("%" "I64u" "\n", a);
	printf("%" PRIu64 "\n", a);

	uint64_t b = 1llu << 63;
	printf("%I64u\n", b);
	printf("%" "I64u" "\n", b);
	printf("%" PRIu64 "\n", b);

	char s[200];
	snprintf(s, 200, "%I64u %I64u", a, b);
	printf("%s\n", s);

	WabtLimits* l =  new WabtLimits;
	l->initial = 100;
	l->max = 100;
	l->has_max = false;

	sprint_limits(s, 200, l);
	puts(s);

	return 0;
}
```

Please note that I did not attempt to override `PRI*64`, all of them are `I64*` as defined in `inttypes.h`. Double checked with the header file and preprocessed output.

Completely bewildered, I then compiled`src\binary-reader.cc`, with and without my workaround, with command flag `g++ -c -D__STDC_FORMAT_MACROS=1 -D__STDC_LIMIT_MACROS=1 src\binary-reader.cc -I src -Wall -Wextra`. Both gave warnings related to `-Wformat`, though warnings were reduced with workaround.

Before giving up, I produced the preprocessed output with and without my workaround then diff them, here is what I get:

```diff
diff --git a/a.cpp b/a.cpp
index 6ff23f1..664af27 100644
--- a/a.cpp
+++ b/a.cpp
@@ -1909,6 +1909,53 @@ extern "C" {
 # 27 "src\\common.h" 2
 
 # 1 "src\\config.h" 1
+# 204 "src\\config.h"
+# 1 "c:\\mingw64\\x86_64-w64-mingw32\\include\\inttypes.h" 1 3
+# 14 "c:\\mingw64\\x86_64-w64-mingw32\\include\\inttypes.h" 3
+# 1 "c:\\mingw64\\lib\\gcc\\x86_64-w64-mingw32\\6.3.0\\include\\stddef.h" 1 3 4
+# 1 "c:\\mingw64\\x86_64-w64-mingw32\\include\\stddef.h" 1 3 4
+# 2 "c:\\mingw64\\lib\\gcc\\x86_64-w64-mingw32\\6.3.0\\include\\stddef.h" 2 3 4
+# 15 "c:\\mingw64\\x86_64-w64-mingw32\\include\\inttypes.h" 2 3
+
+
+extern "C" {
+
+
+typedef struct {
+ intmax_t quot;
+ intmax_t rem;
+ } imaxdiv_t;
+
+
+# 1 "c:\\mingw64\\x86_64-w64-mingw32\\include\\_mingw_print_push.h" 1 3
+# 27 "c:\\mingw64\\x86_64-w64-mingw32\\include\\inttypes.h" 2 3
+# 279 "c:\\mingw64\\x86_64-w64-mingw32\\include\\inttypes.h" 3
+intmax_t __attribute__((__cdecl__)) imaxabs (intmax_t j);
+
+
+
+
+imaxdiv_t __attribute__((__cdecl__)) imaxdiv (intmax_t numer, intmax_t denom);
+
+
+
+intmax_t __attribute__((__cdecl__)) strtoimax (const char* __restrict__ nptr,
+                            char** __restrict__ endptr, int base);
+uintmax_t __attribute__((__cdecl__)) strtoumax (const char* __restrict__ nptr,
+        char** __restrict__ endptr, int base);
+
+intmax_t __attribute__((__cdecl__)) wcstoimax (const wchar_t* __restrict__ nptr,
+                            wchar_t** __restrict__ endptr, int base);
+uintmax_t __attribute__((__cdecl__)) wcstoumax (const wchar_t* __restrict__ nptr,
+        wchar_t** __restrict__ endptr, int base);
+
+
+# 1 "c:\\mingw64\\x86_64-w64-mingw32\\include\\_mingw_print_pop.h" 1 3
+# 300 "c:\\mingw64\\x86_64-w64-mingw32\\include\\inttypes.h" 2 3
+
+
+}
+# 205 "src\\config.h" 2
 # 29 "src\\common.h" 2
 # 71 "src\\common.h"
 
@@ -2628,58 +2675,13 @@ size_t wabt_read_i32_leb128(const uint8_t* ptr,
 
 # 1 "c:\\mingw64\\x86_64-w64-mingw32\\include\\assert.h" 1 3
 # 20 "src\\binary-reader.cc" 2
-# 1 "c:\\mingw64\\x86_64-w64-mingw32\\include\\inttypes.h" 1 3
-# 14 "c:\\mingw64\\x86_64-w64-mingw32\\include\\inttypes.h" 3
-# 1 "c:\\mingw64\\lib\\gcc\\x86_64-w64-mingw32\\6.3.0\\include\\stddef.h" 1 3 4
-# 1 "c:\\mingw64\\x86_64-w64-mingw32\\include\\stddef.h" 1 3 4
-# 2 "c:\\mingw64\\lib\\gcc\\x86_64-w64-mingw32\\6.3.0\\include\\stddef.h" 2 3 4
-# 15 "c:\\mingw64\\x86_64-w64-mingw32\\include\\inttypes.h" 2 3
-
-
-
-# 17 "c:\\mingw64\\x86_64-w64-mingw32\\include\\inttypes.h" 3
-extern "C" {
-
-
-typedef struct {
- intmax_t quot;
- intmax_t rem;
- } imaxdiv_t;
-
-
-# 1 "c:\\mingw64\\x86_64-w64-mingw32\\include\\_mingw_print_push.h" 1 3
-# 27 "c:\\mingw64\\x86_64-w64-mingw32\\include\\inttypes.h" 2 3
-# 279 "c:\\mingw64\\x86_64-w64-mingw32\\include\\inttypes.h" 3
-intmax_t __attribute__((__cdecl__)) imaxabs (intmax_t j);
-
-
 
-
-imaxdiv_t __attribute__((__cdecl__)) imaxdiv (intmax_t numer, intmax_t denom);
-
-
-
-intmax_t __attribute__((__cdecl__)) strtoimax (const char* __restrict__ nptr,
-                            char** __restrict__ endptr, int base);
-uintmax_t __attribute__((__cdecl__)) strtoumax (const char* __restrict__ nptr,
-        char** __restrict__ endptr, int base);
-
-intmax_t __attribute__((__cdecl__)) wcstoimax (const wchar_t* __restrict__ nptr,
-                            wchar_t** __restrict__ endptr, int base);
-uintmax_t __attribute__((__cdecl__)) wcstoumax (const wchar_t* __restrict__ nptr,
-        wchar_t** __restrict__ endptr, int base);
-
-
-# 1 "c:\\mingw64\\x86_64-w64-mingw32\\include\\_mingw_print_pop.h" 1 3
-# 300 "c:\\mingw64\\x86_64-w64-mingw32\\include\\inttypes.h" 2 3
-
-
-}
-# 21 "src\\binary-reader.cc" 2
 # 1 "c:\\mingw64\\x86_64-w64-mingw32\\include\\setjmp.h" 1 3
 # 11 "c:\\mingw64\\x86_64-w64-mingw32\\include\\setjmp.h" 3
 #pragma pack(push,_CRT_PACKING)
 # 26 "c:\\mingw64\\x86_64-w64-mingw32\\include\\setjmp.h" 3
+
+# 26 "c:\\mingw64\\x86_64-w64-mingw32\\include\\setjmp.h" 3
 extern "C" {
 # 112 "c:\\mingw64\\x86_64-w64-mingw32\\include\\setjmp.h" 3
   typedef __attribute__ ((__aligned__ (16))) struct _SETJMP_FLOAT128 {
@@ -3397,22 +3399,10 @@ static WabtResult logging_on_init_expr_get_global_expr(uint32_t value0, uint32_t
 static void sprint_limits(char* dst, size_t size, const WabtLimits* limits) {
   int result;
   if (limits->has_max) {
-    result = snprintf(dst, size, "initial: %" 
-# 643 "src\\binary-reader.cc" 3
-                                                  "llu" 
-# 643 "src\\binary-reader.cc"
-                                                         ", max: %" 
-# 643 "src\\binary-reader.cc" 3
-                                                                    "llu"
-# 643 "src\\binary-reader.cc"
-                                                                          ,
+    result = snprintf(dst, size, "initial: %" "I64u" ", max: %" "I64u",
                            limits->initial, limits->max);
   } else {
-    result = snprintf(dst, size, "initial: %" 
-# 646 "src\\binary-reader.cc" 3
-                                                  "llu"
-# 646 "src\\binary-reader.cc"
-                                                        , limits->initial);
+    result = snprintf(dst, size, "initial: %" "I64u", limits->initial);
   }
   static_cast<void>(result);
   
@@ -3639,11 +3629,7 @@ static WabtResult logging_on_f64_const_expr(uint64_t value_bits,
   LoggingContext* ctx = static_cast<LoggingContext*>(user_data);
   double value;
   memcpy(&value, &value_bits, sizeof(value));
-  do { write_indent(ctx); wabt_writef(ctx->stream, "on_f64_const_expr(%g (0x08%" 
-# 861 "src\\binary-reader.cc" 3
- "llx" 
-# 861 "src\\binary-reader.cc"
- "))\n", value, value_bits); } while (0);
+  do { write_indent(ctx); wabt_writef(ctx->stream, "on_f64_const_expr(%g (0x08%" "I64x" "))\n", value, value_bits); } while (0);
   return ctx->reader->on_f64_const_expr ? ctx->reader->on_f64_const_expr(value_bits, ctx->reader->user_data) : WABT_OK;
 }
 
@@ -3655,15 +3641,7 @@ static WabtResult logging_on_i32_const_expr(uint32_t value, void* user_data) {
 
 static WabtResult logging_on_i64_const_expr(uint64_t value, void* user_data) {
   LoggingContext* ctx = static_cast<LoggingContext*>(user_data);
-  do { write_indent(ctx); wabt_writef(ctx->stream, "on_i64_const_expr(%" 
-# 873 "src\\binary-reader.cc" 3
- "llu" 
-# 873 "src\\binary-reader.cc"
- " (0x%" 
-# 873 "src\\binary-reader.cc" 3
- "llx" 
-# 873 "src\\binary-reader.cc"
- "))\n", value, value); } while (0);
+  do { write_indent(ctx); wabt_writef(ctx->stream, "on_i64_const_expr(%" "I64u" " (0x%" "I64x" "))\n", value, value); } while (0);
   return ctx->reader->on_i64_const_expr ? ctx->reader->on_i64_const_expr(value, ctx->reader->user_data) : WABT_OK;
 }
 
@@ -3762,11 +3740,7 @@ static WabtResult logging_on_init_expr_f64_const_expr(uint32_t index,
   LoggingContext* ctx = static_cast<LoggingContext*>(user_data);
   double value;
   memcpy(&value, &value_bits, sizeof(value));
-  do { write_indent(ctx); wabt_writef(ctx->stream, "on_init_expr_f64_const_expr(index: %u value: %g (0x08%" 
-# 972 "src\\binary-reader.cc" 3
- "llx" 
-# 972 "src\\binary-reader.cc"
- "))\n", index, value, value_bits); } while (0)
+  do { write_indent(ctx); wabt_writef(ctx->stream, "on_init_expr_f64_const_expr(index: %u value: %g (0x08%" "I64x" "))\n", index, value, value_bits); } while (0)
                                 ;
   return ctx->reader->on_init_expr_f64_const_expr ? ctx->reader->on_init_expr_f64_const_expr(index, value_bits, ctx->reader->user_data) : WABT_OK;
 }
@@ -3783,11 +3757,7 @@ static WabtResult logging_on_init_expr_i64_const_expr(uint32_t index,
                                                       uint64_t value,
                                                       void* user_data) {
   LoggingContext* ctx = static_cast<LoggingContext*>(user_data);
-  do { write_indent(ctx); wabt_writef(ctx->stream, "on_init_expr_i64_const_expr(index: %u, value: %" 
-# 989 "src\\binary-reader.cc" 3
- "llu" 
-# 989 "src\\binary-reader.cc"
- ")\n", index, value); } while (0)
+  do { write_indent(ctx); wabt_writef(ctx->stream, "on_init_expr_i64_const_expr(index: %u, value: %" "I64u" ")\n", index, value); } while (0)
              ;
   return ctx->reader->on_init_expr_i64_const_expr ? ctx->reader->on_init_expr_i64_const_expr(index, value, ctx->reader->user_data) : WABT_OK;
 }
```

I am out of idea now, hope my findings will helps.

**Comment by binji** [comment]
OK, was able to repro this on my Windows laptop. Will investigate.

**Comment by binji** [comment]
Believe this is fixed by #346.
